### PR TITLE
[4.0] [New Feature] Cookie Consent

### PR DESCRIPTION
**Pull request by brianteeman**
This plugin uses the cookie consent script from https://cookieconsent.osano.com/ and is a joomla4 version of a plugin I wrote for a non-profit that I support.

## Summary of Changes
This plugin implements a cookie consent notification compliant 

### Why choose Osano Cookie Consent?
- Free and open source.
- Super configurable. Choose your own colors, styles, type of compliance, text and more with a few clicks of the mouse.
- The world’s most popular. Cookie Consent is seen over 2 billion times every month, across millions of websites.

### Informational Consent
You tell your users that you use cookies, and that by continuing to use your website they accept them.

localhost_joomla-cms_index php_en_about(renderforest) (2)-thumb localhost_joomla-cms_index php_en_about(renderforest)-thumb localhost_joomla-cms_index php_en_about(renderforest) (1)-thumb

localhost_joomla-cms_index php_en_about(Pixel 2)-thumb

This is the approach used by most websites, and appears to be sufficient in the UK. Other countries may have stricter rules and may warrant a stricter solutions for which other extensions are available.

https://ico.org.uk/for-organisations/guide-to-pecr/cookies-and-similar-technologies/

It provides your users with no direct control over cookies, other than through their own browser settings.

### What does it look like?
There are 4 different positions
- Top
- Bottom
- Floating left
- Floating right

There are three different layouts
- Block
- Classic
- Edgeless

All colours can be customised in the plugin so you dont need to make any css changes in your template.
- Banner Colour
- Banner Text Colour 
- Button Colour
- Button Text Colour

The text displayed for the message, button and link are customisable in the plugin and translatable with overrides for a multilingual web site

The link is to an article you create to provide more information about cookies on your site. On a multilingual web site this will display the appropriate associated article for that language.

### Consent validity
You can set the expiry date for the consent to
- Never
- 1 week
- 1 month
- 3 months
- 1 year

### Accessibility
Generally speaking this is very good accessible code but it can be improved for multilingual sites and I will submit a pull request upstream for that.

### Default settings
The plugin is not enabled by default. A site owner needs to enable it.

## Test Instructions
- This requires a complete "npm i"
- If its a new install of joomla then just go to the plugins and enable it etc
- If its an existing install you will need to do "database fix" first and then enable the plugin
- Once enabled the cookie consent banner will be enabled on the site with default settings
- Try to test with all the different options
- If you close the banner you will need to clear your cookies if you want to test further.
 
## Todo
- Add documentation to the docs.joomla.org site
- Add post installation message (?)

localhost_joomla-cms_administrator_index php_option=com_plugins view=plugin layout=edit extension_id=207(renderforest)-thumb

### CODE REVIEW

File: plugins/system/cookieconsent/cookieconsent.php [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+ */
+class PlgSystemCookieconsent extends CMSPlugin
+{
+	/**
+	 * If true, language files will be loaded automatically.
+	 *
+	 * @var    boolean
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $autoloadLanguage = true;
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]: Don't need this since `$this->loadLanguage()` is used.

File: plugins/system/cookieconsent/cookieconsent.php [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	/**
+	 * Return the url of the assigned article based on the current user language
+	 *
+	 * @return  string  Returns the link to the cookie policy
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private function getAssignedPolicylinkUrl()
+	{
+		$db = Factory::getDbo();
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]:
```suggestion
		$db = $this->db;
```
And remove `use Joomla\CMS\Factory;`

File: plugins/system/cookieconsent/cookieconsent.php [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+					ContentHelperRoute::getArticleRoute(
+						$article->id,
+						$article->catid,
+						$article->language
+					)
+				);
+			}
+
+			$associatedArticles = Associations::getAssociations('com_content', '#__content', 'com_content.item', $article->id);
+			$currentLang        = Factory::getLanguage()->getTag();
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]:
```suggestion
			$currentLang        = $this->app->getLanguage()->getTag();
```

File: plugins/system/cookieconsent/cookieconsent.php [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+		if ($document->getType() !== 'html')
+		{
+			return;
+		}
+
+		// Load language file.
+		$this->loadLanguage();
+
+		// Get the settings from the plugin
+		$position 			= $this->params->get('position', 'bottom');
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]: Use spaces, not tabs for alignment.

File: plugins/system/cookieconsent/cookieconsent.php [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+			{
+				// Something at the database layer went wrong
+				return Route::_(
+					'index.php?option=com_content&view=article&id='
+					. $this->articleid
+				);
+			}
+
+			// Register ContentHelperRoute
+			JLoader::register('ContentHelperRoute', JPATH_BASE . '/components/com_content/helpers/route.php');
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]: This is not needed.

File: plugins/system/cookieconsent/cookieconsent.php [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		$article = false;
+		$policyArticle = $this->params->get('policylink') > 0 ? (int) $this->params->get('policylink') : 0;
+
+		if ($policyArticle)
+		{
+			// Get the info from the article
+			$query = $db->getQuery(true)
+				->select($db->quoteName(array('id', 'catid', 'language')))
+				->from($db->quoteName('#__content'))
+				->where($db->quoteName('id') . ' = ' . (int) $this->params->get('policylink'));
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]: Use prepared statements.

Comment (brianteeman) [reply plugins/system/cookieconsent/cookieconsent.php]: Please suggest as I don't know how

File: plugins/system/cookieconsent/cookieconsent.php [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+	 * @return  string  Returns the link to the cookie policy
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	private function getAssignedPolicylinkUrl()
+	{
+		$db = Factory::getDbo();
+
+		$article = false;
+		$policyArticle = $this->params->get('policylink') > 0 ? (int) $this->params->get('policylink') : 0;
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]:
```suggestion
		$policyArticle = (int) $this->params->get('policylink');
```

**Comment by brianteeman** [comment]
@SharkyKZ thanks for the review. I will address them all later. Be aware though that almost all the code you commented on is copy paste from other j4 plugins.

File: plugins/system/cookieconsent/cookieconsent.php (hunk 6 above)
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]:
```suggestion
				->where($db->quoteName('id') . ' = :id')
				->bind(':id', $policyArticle, ParameterType::INTEGER);
```
And add `use Joomla\Database\ParameterType;` import.

File: plugins/system/cookieconsent/cookieconsent.php [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+
+		if ($policyArticle)
+		{
+			// Get the info from the article
+			$query = $db->getQuery(true)
+				->select($db->quoteName(array('id', 'catid', 'language')))
```
Comment (SharkyKZ) [inline plugins/system/cookieconsent/cookieconsent.php]:
```suggestion
				->select($db->quoteName(['id', 'catid', 'language']))
```

**Comment by infograf768** [comment]
Suggest to use same system used for offline message, i.e. possibility to use default (core lang  string) or Custom, that last one with the possibility to use a lang constant (i.e. passing through Text::_() )

**Comment by brianteeman** [comment]
> possibility to use default (core lang string) or Custom, that last one with the possibility to use a lang constant (i.e. passing through Text::_() )

@infograf768 it does have that already as stated in the first post. 

image
image

**Comment by infograf768** [comment]
That is not exactly what i suggested.
I suggested to have the choice between a default string which would exist in core and what you display above. similar to what we have for offline message.
this means 2 fields, not one.

**Comment by brianteeman** [comment]
> this means 2 fields, not one.

No thats not a good solution at all and imho should be changed

This is the same solution used for several plugins for a long time so it is consistent behaviour with them.

File: plugins/system/cookieconsent/cookieconsent.xml [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+					<option value="site">PLG_SYSTEM_COOKIECONSENT_SECTION_SITE</option>
+					<option value="administrator">PLG_SYSTEM_COOKIECONSENT_SECTION_ADMIN</option>
+					<option value="both">PLG_SYSTEM_COOKIECONSENT_SECTION_BOTH</option>
+				</field>
+
+				<field
+					name="valid"
+					type="list"
+					default="-1"
+					label="PLG_SYSTEM_COOKIECONSENT_COOKIE_VALID_LABEL"
```
Comment (zero-24) [inline plugins/system/cookieconsent/cookieconsent.xml:38]:
```suggestion
					label="PLG_SYSTEM_COOKIECONSENT_COOKIE_VALID_LABEL"
					validate="options"
```

File: plugins/system/cookieconsent/cookieconsent.xml [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+				/>
+
+				<field
+					name="buttoncolour"
+					type="color"
+					label="PLG_SYSTEM_COOKIECONSENT_BUTTON_COLOUR"
+					default="#ffffff"
+				/>
+
+
```
Comment (zero-24) [inline plugins/system/cookieconsent/cookieconsent.xml]: please remove the extra lines here.

File: administrator/language/en-GB/en-GB.plg_system_cookieconsent.ini [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+PLG_SYSTEM_COOKIECONSENT_POLICY_LABEL="Policy Article"
+PLG_SYSTEM_COOKIECONSENT_POLICY_TEXT_DEFAULT="Learn more"
+PLG_SYSTEM_COOKIECONSENT_POLICY_TEXT_LABEL="Policy Link Text"
+PLG_SYSTEM_COOKIECONSENT_POSITION_LABEL="Position"
+PLG_SYSTEM_COOKIECONSENT_SECTION="Section"
+PLG_SYSTEM_COOKIECONSENT_SECTION_ADMIN="Administrator"
+PLG_SYSTEM_COOKIECONSENT_SECTION_BOTH="Both"
+PLG_SYSTEM_COOKIECONSENT_SECTION_SITE="Site"
+PLG_SYSTEM_COOKIECONSENT_XML_DESCRIPTION="The plugin creates an informational notice to tell your users that you use cookies, and that by continuing to use your website they accept them."
+
```
Comment (zero-24) [inline administrator/language/en-GB/en-GB.plg_system_cookieconsent.ini]: it seams here is an extra line too.

File: administrator/language/en-GB/en-GB.plg_system_cookieconsent.ini [hunk 12]
```diff
@@ -0,0 +1,40 @@
+; Joomla! Project
+; Copyright (C) 2005 - 2019 Open Source Matters. All rights reserved.
+; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
+; Note : All ini files need to be saved as UTF-8
+
+
```
Comment (zero-24) [inline administrator/language/en-GB/en-GB.plg_system_cookieconsent.ini]: please remove the extra lines here.

File: administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-09-14.sql [hunk 13]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO `#__extensions`
```
Comment (zero-24) [inline administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-09-14.sql]: can we use this notation for the update sql: https://github.com/joomla/joomla-cms/blob/566e8b03d353064d89b0c45951a5e49e6aed7969/administrator/components/com_admin/sql/updates/mysql/4.0.0-2019-03-31.sql

Means we have one insert into line including the fields and values at the end and that the actually insert into an dedicated line.

File: administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-09-14.sql [hunk 14]
```diff
@@ -0,0 +1,4 @@
+INSERT INTO "#__extensions"
```
Comment (zero-24) [inline administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-09-14.sql]: same as for the other sql ;)

**Comment by infograf768** [comment]
>No thats not a good solution at all and imho should be changed

> This is the same solution used for several plugins for a long time so it is consistent behaviour with them.

A few remarks and bugs

How does the user know that he has **to create a totally new lang string constant** in override as overriding the default

`PLG_SYSTEM_COOKIECONSENT_MESSAGE_TEXT_DEFAULT` by
creating an override for the default
`PLG_SYSTEM_COOKIECONSENT_MESSAGE_TEXT_DEFAULT="This website uses cookies to ensure you get the best experience on our website. If you do not know you can create an override and create it for administrator and site, hard luck for you.\n\nAlso, sorry, but Clicking on 'Learn More' will not work because the administrator of this site have not chosen an article to display."`

=> No message display in frontend. Bug imho.
Curiously, it displays fine when editing the plugin:
Screen Shot 2019-09-15 at 10 01 39


Let's say this is corrected: How to know this without very complex explanations?
No tip to explain that. Again something hidden in a doc somewhere?

### Other test, article
Keeping the default without override:
If one does not create an article Learn More is still present. Bug.
Create an article for the Learn More:
It displays in the component area instead of a modal. I suggest to modify ths behavior.

### Other test: new constant 
Creating a new constant for the Message:
`PLG_SYSTEM_COOKIECONSENT_MESSAGE_MYTEXT="This website uses cookies to ensure you get the best experience on our website. If you do not know you can create an override and create it for administrator, hard luck for you.\n\nAlso, sorry, but Clicking on 'Learn More' will not work because the administrator of this site have not chosen an article to display."`

Screen Shot 2019-09-15 at 09 18 21

Override string
Screen Shot 2019-09-15 at 09 20 54

Result => untranslated string
Screen Shot 2019-09-15 at 09 22 39

### Testing further
I added the string in en-GB.ini.
No change. =>bug.

basically, the only way I found to modify the message was to edit the string `PLG_SYSTEM_COOKIECONSENT_MESSAGE_TEXT_DEFAULT` in the `en-GB.plg_system_cookieconsent.ini` file or enter a text in the field.

### Learn More
This one can be overriden fine
`PLG_SYSTEM_COOKIECONSENT_POLICY_TEXT_DEFAULT="Learn more if you need"`
works ok if you add in site language overrides.

Hope this helps.

**Comment by brianteeman** [comment]
I will repeat - the code/ui here is the same as for the user/terms plugin, the content/confirmconsent plugin and the user/profile plugin

**Comment by infograf768** [comment]
> I will repeat - the code/ui here is the same as for the user/terms plugin, the content/confirmconsent plugin and the user/profile plugin

I have just demonstrated that it is broken for this one.
Just test.

**Comment by brianteeman** [comment]
> PLG_SYSTEM_COOKIECONSENT_MESSAGE_TEXT_DEFAULT in the en-GB.plg_system_cookieconsent.ini file or enter a text in the field.

That is the expected behaviour

**Comment by infograf768** [comment]
> That is the expected behaviour

I guess you are very confused. Read your own post above:
https://github.com/joomla/joomla-cms/pull/26301#issuecomment-531499882

**Comment by infograf768** [comment]
Found out the culprit for the message override.
The override created value should never contain non-escaped singlequotes or counterslashes.

**Comment by infograf768** [comment]
Or. rather, adding addslashes().
Example in `.../plugins/system/cookieconsent/cookieconsent.php`
`$message          = addslashes($this->params->get('message-text', Text::_('PLG_SYSTEM_COOKIECONSENT_MESSAGE_TEXT_DEFAULT')));`

will solve the issue.
same for other possible overrides

**Comment by infograf768** [comment]
Which gives (to also take care of the absence of an article):

```
		// Get the settings from the plugin
		$position         = $this->params->get('position', 'bottom');
		$layout           = $this->params->get('layout', 'block');
		$bannercolour     = $this->params->get('bannercolour', '#000000');
		$buttoncolour     = $this->params->get('buttoncolour', '#ffffff');
		$buttontextcolour = $this->params->get('buttontextcolour', '#383b75');
		$bannertextcolour = $this->params->get('bannertextcolour', '#f1d600');
		$message          = addslashes($this->params->get('message-text', Text::_('PLG_SYSTEM_COOKIECONSENT_MESSAGE_TEXT_DEFAULT')));
		$dismiss          = addslashes($this->params->get('button-text', Text::_('PLG_SYSTEM_COOKIECONSENT_BUTTON_TEXT_DEFAULT')));
		$valid            = $this->params->get('valid', '-1');
		$href             = $this->getAssignedPolicylinkUrl();
		$link             = $href ? addslashes($this->params->get('policylink-text', Text::_('PLG_SYSTEM_COOKIECONSENT_POLICY_TEXT_DEFAULT'))) : '';

```

**Comment by infograf768** [comment]
I have tested this item :red_circle: unsuccessfully on 222c2ea0764da43e6b7bbfe0927a8d757e81de30<br /><br />Missing addslashes as posted above.

Also, contrary to Terms and Conditions for example,the article is not loaded in a modal.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26301">issues.joomla.org/tracker/joomla-cms/26301</a>.</sub>

**Comment by unknown-sudo** [comment]
The Cookie Directive is changing to the effect that the previously used opt-out variant is then no longer permitted. With the opt-out variant, the visitor does not necessarily have to give consent to the use of cookies, but tacit consent was possible.

This is now changing to the effect that consent to the use of cookies is required or, alternatively, the deactivation of cookies should be possible. All this under the premise that the website content should also be made available if the visitor deactivates the use of cookies and if possible directly at the beginning before the actual use of the website. 

In addition the possibility to maintain the external tracking scripts is missing here. If you only use Joomla session cookies you don't need the consent anyway.

**Comment by brianteeman** [comment]
> The Cookie Directive is changing to the effect that the previously used opt-out variant is then no longer permitted. 

I didnt find anything about this change but its not relevant to this pr anyway

> In addition the possibility to maintain the external tracking scripts is missing here. 

Beyond the scope of this pr as described in the original post and extensions are available for this

> If you only use Joomla session cookies you don't need the consent anyway.

Its not just session cookies that joomla sets ;) I personally agree with you that they are cookies that you dont need consent for but it seems that the majority of the uninformed do not agree.

**Comment by HLeithner** [comment]
> > The Cookie Directive is changing to the effect that the previously used opt-out variant is then no longer permitted.
> 
> I didnt find anything about this change but its not relevant to this pr anyway

I think he is talking about this [german article](https://www.golem.de/news/manipulierte-zustimmung-die-meisten-cookie-banner-sind-illegal-1909-143773.html) and the paper [(Un)informed Consent: Studying GDPR Consent Notices in the Field](https://www.syssec.ruhr-uni-bochum.de/media/emma/veroeffentlichungen/2019/09/05/uninformed-consent_Yl7FPEh.pdf) it's based on

**Comment by brianteeman** [comment]
Well then that doesn't apply here as it is not about GDPR and I prefer to work with what the law says not what an article says

**Comment by HLeithner** [comment]
> Well then that doesn't apply here as it is not about GDPR and I prefer to work with what the law says not what an article says

That's what they do too ;-)

> On May 25, 2018, the European Union’s General Data Protection
Regulation (GDPR; Regulation (EU) 2016/679) went into effect. Its
Article 6 contains six legal bases for the processing of personal
data of European residents, including that “the data subject has
given consent to the processing of his or her personal data for one
or more specific purposes”. Recital 32 of the GDPR and guidelines
published by EU data protection authorities [6] require for valid
consent “a clear affirmative act” that is a “freely given, [purpose-
]specific, informed and unambiguous indication of [...] agreement 
to the processing of personal data.” Another document clarifies the
relationship between the ePrivacy Directive (2002/58/EC) and the
GDPR for the use of cookies: Article 5(3) of the directive governs
access to non-necessary cookies in the user’s browser, whether it
contains personal data or not, while the GDPR applies to subsequent
processing of personal data retrieved via cookies [15].

And as you mention already in your description:
> This is the approach used by most websites, and appears to be sufficient in the UK. Other countries may have stricter rules and may warrant a stricter solutions for which other extensions are available.

I don't think it makes sense to only full fill the requirements to the subset of the countries requiring the cookie policy. Also the ePrivacy directive could be more strict (if it ever comes).

**Comment by brianteeman** [comment]
This is not about GDPR and personal data

**Comment by blueforce** [comment]
patch could not be applied for test...
Bildschirmfoto 2019-10-19 um 12 25 17

**Comment by oschuldt** [comment]
some basic thoughts about cookie consent:
actually it is about gdpr according to Recital 78 - privacy by default (https://gdpr-info.eu/recitals/no-78/ ). But foremost it is about the Directive 2002/58/EC - Article 5 3. wich says: within the EU and for EU citizens, an explicit opt-in procedure must be used. However, this applies exclusively to the use of tracking cookies and cookies for advertising purposes.
Session cookies, user cookies or 'remember me' cookies do not need any consent or information.

I'm afraid that this plugin is not necessary in this form, and may even cause damage if someone thinks it is legally compliant.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26301">issues.joomla.org/tracker/joomla-cms/26301</a>.</sub>

**Comment by oschuldt** [comment]
I have tested this item :white_check_mark: successfully on 55532f0e0e78915d09e8783d8b97c81a8125d1d6<br /><br />works but is not sufficient to run a gdpr/ePrivacy compliant Website! (see comment)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26301">issues.joomla.org/tracker/joomla-cms/26301</a>.</sub>

**Comment by nadjak77** [comment]
I have tested this item :red_circle: unsuccessfully on 55532f0e0e78915d09e8783d8b97c81a8125d1d6<br /><br />cannot test, npm i error (Cannot find module 'walk-sync')<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/26301">issues.joomla.org/tracker/joomla-cms/26301</a>.</sub>

**Comment by tecpromotion** [comment]
> Session cookies, user cookies or 'remember me' cookies do not need any consent or information.

With this information and the hint that you may treat tracking and further analysis cookies separately and may not set cookies without consent, the question arises whether this should really be a core plugin.
However, the "Cookie Consent by Osano" script can do even more. It offers the possibilities of consent and rejection. But also not without adjustments by the user and then it again becomes very complicated for the normal user. But there are already many extensions in JED that do this partly based on the "Osano" solution or even go further than required today.
